### ISSUE TITLE
Add support for CDNs

### ISSUE THREAD

**Issue by dziudek** [GetPublii/Publii#274]
ref. https://github.com/GetPublii/Publii/issues/271#issuecomment-454391795

We need to add an option to specify a domain for the assets (it would be good to specify separate domains for theme assets and media assets).

**Comment by sciita** [comment]
No, the relative path needed to enable netlify cdn.

And I am using cdn for static assets (fonts, images, svgs, js, css ). And I don't enable them manually, it's automatic in netlify ( if you choose proper settings ).

**Comment by docherty** [comment]
Here is the Ghost storage adapter I've used for Cloudinary which worked perfectly for my needs. Perhaps there's some value to be had looking at that as a starting point.
https://github.com/eexit/ghost-storage-cloudinary

**Comment by designgrande** [comment]
When can we expect support for CDN?

**OR alternatively** can developers please provide or tell me how to modify static assets (js/CSS/fonts/images) URL path to custom CDN path.

Example: After sites rendering; the path of a CSS file is:
domain.com/assets/css/style.css

To enable CDN, I need to change it to:
cdn.domain.com/assets/css/style.css

BUT it is necessary that there is "style.css" in the following directory -> domain.com/assets/css/

So basically, I just need help to rewrite the URL for static assets after or while rendering.

I want to transfer 3 of my WordPress websites to publii, but I can not afford to sacrifice on website speed, and without CDN, I don't find any mean to keep all my site performant.

Please please help me with this.

Thanks!

**Comment by designgrande** [comment]
Any help, please

**Comment by dziudek** [comment]
@designgrande - at this moment the only way to make it working is to use `@rendering.previewMode` global variable to create custom CSS/JS paths when website is rendered. So if previewMode is active - you render CSS/JS paths in the old way, but for the non-preview mode you use custom paths.

**Comment by designgrande** [comment]
@dziudek Hi, I understood how to do it for CSS and JS (also thanks to docs: https://getpublii.com/dev/renderer-global-variable/) but how to achieve the same with images?

**Comment by gadgetraja** [comment]
Any update on CDN implementation - I think it was requested back in Jan 2019. Not enforcing or anything, just that every other issue is resolved within months.

Regards.

**Comment by Suleman-Elahi** [comment]
Is there any improvement on this. Because SEO site checkers give warnings about it. I not images then at least there should be an option to load most common CSS/JS files from a CDN.

**Comment by thefourcraft** [comment]
I think that this problem is still, applicable to this day. If I want to add my own domain to GitHub pages I need to change all the files manually. 
adding a check box for a domain will be amazing when pushing to Github pages because it will mess up the website

it won't find the CSS...

**Comment by candideu** [comment]
> Here is the Ghost storage adapter I've used for Cloudinary which worked perfectly for my needs. Perhaps there's some value to be had looking at that as a starting point. https://github.com/eexit/ghost-storage-cloudinary

Ghost also has a an official Cloudinary integration: https://ghost.org/integrations/cloudinary/

**Comment by bobmitro** [comment]
Since Publii now supports plugins, this can be handled with a dedicated plugin. However, it can also be achieved using the existing [Smart Replacer,](https://marketplace.getpublii.com/plugins/smart-replacer/) plugin which allows you to replace image URLs with those from a CDN.

**Comment by bobmitro** [comment]
Closed: This is now part of Publii [v. 0.46](https://getpublii.com/blog/release-046.html)

**Comment by vivek-kumar-poddar** [comment]
So for a basic functionality, which by the way is crucial for any website that cares about speed and usability, I have to pay $59 annually and I do not even get a CDN. Nice!

**Comment by bobmitro** [comment]
What do you have to pay for?

**Comment by vivek-kumar-poddar** [comment]
> What do you have to pay for?

"Since Publii now supports plugins, this can be handled with a dedicated plugin. However, it can also be achieved using the existing [Smart Replacer,](https://marketplace.getpublii.com/plugins/smart-replacer/) plugin which allows you to replace image URLs with those from a CDN."

Smart Replacer - €59.00 - 12-months access to the files and updates - 3-months premium support

You yourself pointed it

**Comment by dziudek** [comment]
@vivek-kumar-poddar

> this can be handled with a dedicated plugin

you can always check the plugin docs, check existing ones and create a simple plugin which will fit your needs: https://github.com/GetPublii/Publii/discussions/1359

**Comment by bobmitro** [comment]
You can purchase it, but you don’t have to - this is my suggestion for solving your problem. Publii is open-source software, so you can write the extension yourself or ask someone else to do it.  

At this moment, we are not planning to create such a plugin because we have other priorities. We might consider it in the future, but not right now.
